File: mcp/tasks/stock/price.py
import logging
import random
from datetime import datetime

logger = logging.getLogger(__name__)

class StockPrice:
    """Task that fetches the current price of a stock by its ticker symbol"""
    
    description = "Get the current price of a stock by ticker symbol"
    
    input_schema = {
        "type": "object",
        "properties": {
            "ticker": {
                "type": "string",
                "description": "The ticker symbol of the stock"
            }
        },
        "required": ["ticker"]
    }
    
    output_schema = {
        "type": "object",
        "properties": {
            "ticker": {
                "type": "string",
                "description": "The ticker symbol of the stock"
            },
            "price": {
                "type": "number",
                "description": "The current price of the stock"
            },
            "currency": {
                "type": "string",
                "description": "The currency of the price"
            },
            "timestamp": {
                "type": "string",
                "description": "The timestamp of the price"
            }
        },
        "required": ["ticker", "price", "currency", "timestamp"]
    }
    
    @staticmethod
    def run(input_data):
        """Run the stock price task"""
        ticker = input_data.get("ticker")
        
        if not ticker:
            raise ValueError("Ticker symbol is required")
            
        # In a real implementation, this would call a stock API
        # For now, we'll just return mock data
        logger.info(f"Fetching stock price for {ticker}")
        
        # Create a deterministic but random-looking price based on ticker
        seed = sum(ord(c) for c in ticker)
        random.seed(seed)
        price = round(random.uniform(10, 1000), 2)
        
        return {
            "ticker": ticker,
            "price": price,
            "currency": "USD",
            "timestamp": datetime.now().isoformat()
        }
        
# Export the task instance
stock_price = StockPrice

File: src/tasks/stock/price.ts
import { ForgeTask } from '@forgehive/forge-runner';
import axios from 'axios';

interface StockPriceInput {
  ticker: string;
}

interface StockPriceOutput {
  ticker: string;
  price: number;
  currency: string;
  timestamp: string;
}

/**
 * Task that fetches the current price of a stock by its ticker symbol
 */
export const stockPrice: ForgeTask<StockPriceInput, StockPriceOutput> = {
  name: 'stock:price',
  description: 'Get the current price of a stock by its ticker symbol',
  
  // Define the input schema for this task
  inputSchema: {
    type: 'object',
    properties: {
      ticker: {
        type: 'string',
        description: 'The stock ticker symbol (e.g., AAPL for Apple Inc.)',
      },
    },
    required: ['ticker'],
  },
  
  // Define the output schema for this task
  outputSchema: {
    type: 'object',
    properties: {
      ticker: {
        type: 'string',
        description: 'The stock ticker symbol',
      },
      price: {
        type: 'number',
        description: 'The current stock price',
      },
      currency: {
        type: 'string',
        description: 'The currency of the price (e.g., USD)',
      },
      timestamp: {
        type: 'string',
        description: 'The timestamp when the price was fetched',
      },
    },
  },
  
  // Implement the actual task execution
  async run({ ticker }) {
    try {
      // In a real implementation, you would use a proper financial API
      // For this example, we're simulating the API call
      // Replace this with an actual API call in production
      
      // For demonstration, using a free API (you should use a more reliable API in production)
      const apiKey = process.env.STOCK_API_KEY || 'demo';
      const response = await axios.get(
        `https://www.alphavantage.co/query?function=GLOBAL_QUOTE&symbol=${ticker}&apikey=${apiKey}`
      );
      
      // Parse the response
      const data = response.data;
      
      // Check if we have valid data
      if (!data || !data['Global Quote'] || !data['Global Quote']['05. price']) {
        throw new Error(`Unable to fetch price for ticker ${ticker}`);
      }
      
      const price = parseFloat(data['Global Quote']['05. price']);
      
      return {
        ticker: ticker.toUpperCase(),
        price,
        currency: 'USD', // Assuming USD for simplicity
        timestamp: new Date().toISOString(),
      };
    } catch (error) {
      console.error(`Error fetching stock price for ${ticker}:`, error);
      throw new Error(`Failed to fetch stock price for ${ticker}: ${error.message}`);
    }
  },
};
